pysrc/bytewax/connectors/kafka/registry.py

```diff
@@ -0,0 +1,222 @@
+"""TODO."""
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+import requests
+from confluent_kafka.schema_registry import SchemaRegistryClient
+
+from .serde import (
+    SchemaDeserializer,
+    SchemaSerializer,
+    _AvroDeserializer,
+    _AvroSerializer,
+    _ConfluentAvroDeserializer,
+    _ConfluentAvroSerializer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class SchemaConf:
+    """Info used to retrieve a schema from a schema registry.
+
+    Either use `schema_id`, or specify the `subject` and optionally `version`.
+    If no `version` is specified, it defaults to the latest schema.
+    """
+
+    # TODO: Only `avro` supported for now, but we might want
+    #       to add `protobuf` and others too
+    # format: str = "avro"
+    schema_id: Optional[int] = None
+    subject: Optional[str] = None
+    version: Optional[int] = None
+
+    def __post_init__(self):
+        """Validate the data."""
+        if self.schema_id is None:
+            if self.subject is None:
+                msg = "subject MUST be specified if no schema_id is provided"
+                raise ValueError(msg)
+        else:
+            if self.subject is not None:
+                logger.warning(
+                    "both schema_id and subject specified, subject will be ignored"
+                )
+
+
+class SchemaRegistry(ABC):
+    """An interface to define a SchemaRegistry.
+
+    A schema registry must define functions to build
+    a (de)serializer for both key and value of a KafkaMessage.
+    """
+
+    @abstractmethod
+    def key_serializer(self) -> SchemaSerializer:
+        """Return a serializer for keys."""
+        ...
+
+    @abstractmethod
+    def value_serializer(self) -> SchemaSerializer:
+        """Return a serializer for values."""
+        ...
+
+    @abstractmethod
+    def key_deserializer(self) -> SchemaDeserializer:
+        """Return a deserializer for keys."""
+        ...
+
+    @abstractmethod
+    def value_deserializer(self) -> SchemaDeserializer:
+        """Return a deserializer for values."""
+        ...
+
+
+class ConfluentSchemaRegistry(SchemaRegistry):
+    """Confluent's schema registry for Kafka's input and output connectors.
+
+    Serializer: Use either `schema_id` or `subject`+`version` to fetch
+    the schema for the serializer from the registry.
+
+    Deserializer: The deserializer automatically fetches the correct
+    schema for each message, following confluent_kafka's behaviour
+    """
+
+    def __init__(
+        self,
+        sr_conf: Dict,
+        key_serialization_schema: Optional[SchemaConf] = None,
+        value_serialization_schema: Optional[SchemaConf] = None,
+        enable_key_deserialization: bool = True,
+        enable_value_deserialization: bool = True,
```

Comment:
It's in the docstrings. Confluent's schema registry doesn't need a schema specification to deserialize messages, since its client will read the magic bytes and always fetch the right schema. So you only decide if you want deserialization or not, not which schema you can use. Serialization instead requires to specify a schema.

Redpanda doesn't have the info to infer the schema of a message that has to be deserialized, so you have to specify that too.

> What's your thinking behind defining this on the SchemaRegistry itself and not where it is used? Configuring "here's how I connect to my schema registry" is a different question than "here's the schemas this specific input should use".

The point is that depending on the use case, one might want to only deserialize the `value` of an incoming message, and that's it. Or deserialize both value and key, then only serialize the value, or all the possible combinations. Here a single schema registry instance is able to handle all the combinations. Then it creates the specific (de)serializers you might need. But I agree that the separating the two things might lead to a better api, I'll work on this.